src/main/java/net/neoforged/neoforge/registries/IRegistryExtension.java

```diff
@@ -0,0 +1,87 @@
+package net.neoforged.neoforge.registries;
+
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.registries.callback.RegistryCallback;
+
+public interface IRegistryExtension<T> {
+    /**
+     * {@return whether this registry should be synced to clients}
+     */
+    boolean doesSync();
+
+    /**
+     * {@return the highest id that an entry in this registry is <i>allowed</i> to use}
+     * For the size of this registry, see {@link Registry#size()}.
+     */
+    int getMaxId();
+
+    /**
+     * Adds a callback to this registry.
+     * <p>
+     * The callback will be called when the registry is added to, baked, or cleared.
+     *
+     * @param callback The callback to add.
+     */
+    void addCallback(RegistryCallback<T> callback);
+
+    default <C extends RegistryCallback<T>> void addCallback(Class<C> type, C callback) {
```

Comment:
So that you can do `addCallback(AddCallback.class, (....) -> )` otherwise you need an explicit cast as the `RegistryCallback` variant is not type explicit.

Reply:
Why do we need a top-level empty `RegistryCallback` interface? It would be better to use 3 methods, one for each callback.

Reply:
Because otherwise having the same callback object implement multiple times would need an explicit registration as a `XCallback` every time

Reply:
Ah, interesting, I had missed that. Can you document this behavior on the first method?

I think this second method is unfortunately not idiomatic java, and we should probably let people write an explicit cast, i.e. `addCallback((AddCallback) (...) -> ...)`, instead.

Reply:
Added a bit of javadoc in d2d5141.